crypto/interface.go

```diff
@@ -34,28 +37,39 @@ type KeyCreator interface {
 
 // KeyResolver defines the functions for retrieving keys.
 type KeyResolver interface {
-	// GetPublicKey returns the PublicKey
-	// If a key is missing, a Storage.ErrNotFound is returned
-	GetPublicKey(kid string) (crypto.PublicKey, error)
+	// GetPublicKey returns the PublicKey if it was valid on the given validationTime
+	// If a key is missing, a Storage.ErrNotFound is returned.
+	GetPublicKey(kid string, validationTime time.Time) (crypto.PublicKey, error)
+}
+
+// PublicKeyStore defines the functions for retrieving and storing public keys.
+type PublicKeyStore interface {
+	KeyResolver
+
+	// SavePublicKey stores or overwrites a public key with a given kid.
+	// The validity period defines when the given key is valid.
+	SavePublicKey(kid string, publicKey crypto.PublicKey, period core.Period) error
 }
 
 // KeyStore defines the functions that can be called by a Cmd, Direct or via rest call.
 type KeyStore interface {
 	KeyCreator
-	KeyResolver
-	JWSSigner
-	// GetPrivateKey returns the specified private key (for e.g. signing) in non-exportable form.
-	// If a key is missing, a Storage.ErrNotFound is returned
-	GetPrivateKey(kid string) (crypto.Signer, error)
-	// SignJWT creates a signed JWT using the given key and map of claims (private key must be present).
-	SignJWT(claims map[string]interface{}, kid string) (string, error)
+	PublicKeyStore
+	DocSigner
+	AuthSigner
 	// PrivateKeyExists returns if the specified private key exists.
 	// If an error occurs, false is also returned
 	PrivateKeyExists(kid string) bool
 }
 
-// JWSSigner defines the functions for signing JSON Web Signatures.
-type JWSSigner interface {
+// DocSigner defines the functions for signing JSON Web Signatures.
+type DocSigner interface {
```

Comment:
not sure if  Document in `DocSigner` is the correct term, since that feels like it's borrowed from RFC004 or RFC006, which means where a bit of mixing between layers. Can't it share the same interface (maybe `JOSESigner`?) with `SignJWT`?

Reply:
Used old name. Merging both signing interfaces gave quite some work in some tests